File: cpp/cmd/basename.cpp
// SPDX-License-Identifier: MIT
#include <koru/braam.hpp>

namespace {

constexpr Str USAGE =
    "Usage:\n"
    "    basename <path> [<suffix>]\n"
    "    basename [-a] [-s <suffix>] <path>...\n"
    "Options:\n"
    "    -a    every operand is a path, none of them a suffix\n"
    "    -s    strip this suffix, and imply -a\n";

// Text, not a path: nothing here opens anything. path.cpp's path_basename
// takes a normalised absolute path, so the trailing slashes go first.
Str base_of(Str p, Str suffix)
{
    if (p.empty())
        return p;
    while (p.size() > 1 && p[p.size() - 1] == '/')
        p = p.substr(0, p.size() - 1);
    if (p == "/")
        return p;

    Str name = path_basename(p);
    // A name that is only the suffix keeps it, rather than coming out empty.
    if (!suffix.empty() && name != suffix && name.ends_with(suffix))
        name = name.substr(0, name.size() - suffix.size());
    return name;
}

} // namespace

Task<i32> proc_main(Args args)
{
    bool all = false;
    Str suffix;

    if (args.size() == 1 || help_asked(args))
        co_return co_await usage_asked(USAGE);

    OptParse p(args, Opts{ "a", "s" });
    for (Opt o;;) {
        Result<bool> r = p.next(o);
        if (r.is_err()) {
            Buf<64> b;
            b.put("basename: ")
                .put(r.error() == Error::NotFound ? "option requires an argument -- "
                                                  : "illegal option -- ")
                .put(o.name)
                .put('\n');
            co_await write_all(SYS_STDERR, b.str());
            co_return co_await usage_error(USAGE);
        }
        if (!r.value())
            break;
        if (o.name == 's')
            suffix = o.value;
        all = true; // -s implies -a
    }

    // Without either flag a second operand is the suffix, and there is no third.
    Args rest  = p.rest();
    usize take = rest.size();
    if (take == 0 || (!all && take > 2))
        co_return co_await usage_error(USAGE);
    if (!all && take == 2) {
        suffix = rest[1];
        take   = 1;
    }

    String out;
    for (usize i = 0; i < take; i++)
        if (!out.append(base_of(rest[i], suffix)) || !out.push('\n'))
            co_return 1;
    if ((co_await write_all(SYS_STDOUT, out.str())).is_err())
        co_return 1;
    co_return 0;
}

File: rust/runtime/examples/basename.rs
// SPDX-License-Identifier: MIT

//! Braam's `src/cmd/basename.cpp`, in Rust. The option parser, the path
//! helpers and nothing else: no ring call but the one write at the end.

use koru::{Args, Kind, OptParse, Opts, Str, help_asked, usage_asked, usage_error, write_all};

const USAGE: &str = "Usage:
    basename <path> [<suffix>]
    basename [-a] [-s <suffix>] <path>...
Options:
    -a    every operand is a path, none of them a suffix
    -s    strip this suffix, and imply -a
";

/// Text, not a path: nothing here opens anything. A name that is only the
/// suffix keeps it, rather than coming out empty.
fn base_of<'a>(mut p: Str<'a>, suffix: Str<'_>) -> Str<'a> {
    if p.is_empty() {
        return p;
    }
    while p.len() > 1 && p.ends_with('/') {
        p = &p[..p.len() - 1];
    }
    if p == "/" {
        return p;
    }

    let name = match p.rfind('/') {
        Some(i) => &p[i + 1..],
        None => p,
    };
    if !suffix.is_empty() && name != suffix && name.ends_with(suffix) {
        return &name[..name.len() - suffix.len()];
    }
    name
}

#[koru::main]
async fn main(args: Args) -> i32 {
    if args.size() == 1 || help_asked(&args) {
        return usage_asked(USAGE).await;
    }

    let mut all = false;
    let mut suffix = "";
    let mut p = OptParse::new(
        &args,
        Opts {
            flags: "a",
            valued: "s",
        },
    );
    loop {
        match p.next() {
            Err(e) => {
                let why = if e.error.is(Kind::NotFound) {
                    "option requires an argument -- "
                } else {
                    "illegal option -- "
                };
                let line = format!("basename: {}{}\n", why, e.name);
                write_all(koru::stderr(), &line).await.ok();
                return usage_error(USAGE).await;
            }
            Ok(None) => break,
            Ok(Some(o)) => {
                if o.name == 's' {
                    suffix = o.value;
                }
                all = true; // -s implies -a
            }
        }
    }

    // Without either flag a second operand is the suffix, and there is no third.
    let rest = p.rest();
    let mut take = rest.size();
    if take == 0 || (!all && take > 2) {
        return usage_error(USAGE).await;
    }
    if !all && take == 2 {
        suffix = &rest[1];
        take = 1;
    }

    let mut out = String::new();
    for i in 0..take {
        out.push_str(base_of(&rest[i], suffix));
        out.push('\n');
    }
    i32::from(write_all(koru::stdout(), &out).await.is_err())
}
